determined-ai/determined
# feat: add auto create namespace

## Ticket
DET-10224


## Description
Implement backend logic for auto-creating Kubernetes namespaces and implement the CLI arguments for this feature.
`det w create <workspace-name> --cluster-name <cluster_name> --auto-create-namespace`,
 `det w create <workspace_name> --auto-create-namespace-all-clusters`, 
`det w bindings set  <workspace-name> --cluster-name <cluster_name> --auto-create-namespace`, and 
` det w bindings set <workspace-name>  --auto-create-namespace-all-clusters`

## Test Plan
- Start up a multiRM cluster using two kubernetes resource managers with names `c1` and `c2` respectively.
- Run `det w create <workspace-name> --cluster-name c1 --auto-create-namespace` 
- Run `kubectl get namespaces` and verify that the namespace name returned from the above CLI command is listed
- Run `det w bindings list` and verify that the correct workspace-namespace bindings show up (`default` for `c2`, and the auto-created namespace name for `c1` 

Now,  create a workspace and verify that the auto-create feature works with `det w bindings set` 
- Run `det w create <second_workspace_name>`
- Run `det w bindings set <second_workspace_name> --auto-create-namespace --cluster_name c2`
- Run `det w bindings list` and verify that the auto-created namespace is listed for cluster `c1` while the default namespace is listed for cluster `c2` 

- Run `det w create <third_workspace_name> --auto-create-namespace-all-clusters` and verify that the output of this command shows workspace-namespace bindings to an auto-created namespace for both clusters.

- Run `det w bindings list` and verify that both auto-created namespaces show up appropriately.

 
-Run `det w create <fourth_workspace_name>` and `det w bindings set <fourth_workspace_name> --auto-create-namespace-all-clusters` and verify the output shows the auto-created workspace-namespace bindings. 
- Run `det w bindings list <fourth_workspace_name>` to verify that both auto-created namespaces show up appropriately.

## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.




[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
